ed25519.go

```diff
@@ -157,6 +157,12 @@ func Sign(privateKey PrivateKey, message []byte) []byte {
 	// S = (r + H(R,A,m)a) mod L
 	modm.Contract(RS[32:], &S)
 
+	h.Reset()
+	a.Reset()
+	for i := range extsk {
+		extsk[i] = 0
```

Comment:
The go compiler, even if it doesn't now, could be extended to learn that this is a dead store: `extsk` is going out of scope, so dead store elimination would remove this memory cleansing step.  Of course, golang is a garbage collected language -- and there are no guarantees, AFAIK, that it isn't a copying GC -- so that a GC run before the zeroization could have left an old copy containing sensitive values due to compaction copying or generation promotion, etc.

That is not to say we shouldn't do this necessarily, but just that it's not possible to do a perfect job given the language semantics.  Looking at the existence of the `unsafe` package, it may be the case that once an object pointer has been converted to an unsafe pointer, the GC system cannot move it -- as long as it's subject to the usual lifetime rules, this is not a huge burden.  So there may be a way to allocate `extsk`, create an unsafe pointer to prevent GC moves/copies, and then do the zeroization (possibly having to call out to non-golang code?).  I haven't researched this and haven't found a lot discussions about this sort of stuff though.

Reply:
Yeah, this is a "best-effort" type thing, that happens to work for the current implementation.

There's a some `unsafe` trickery that can be used to call the runtime's assembly optimized (opaque to the optimizer) `bzero` equivalent (`runtime.memclrNoHeapPointers`), but the compiler is doing that for me under the hood already (as an optimization) so I didn't bother.